Fixes for Clocker

grkvlt opened the pull request:
Some minor fixes to the Docker provider discovered running Clocker.

- The `DockerTemplateOptions` code needed tidying and incosistencies fixed
- The OkHttp connection should support plaintext as well as TLS, to allow Powerstrip proxying
- The `HostConfig` domain object should support a list of `Dns` strings

grkvlt commented:
@andreaturli Does this look OK? What about backporting to the 1.9.x branch?

andreaturli commented:
Lgtm

nacx commented on docker/src/main/java/org/jclouds/docker/compute/options/DockerTemplateOptions.java:
```diff
@@ -384,7 +400,7 @@ public static DockerTemplateOptions overrideLoginCredentials(LoginCredentials cr
+       * @see TemplateOptions#blockUntilRunning(boolean)
        */
       public static DockerTemplateOptions blockUntilRunning(boolean blockUntilRunning) {
          DockerTemplateOptions options = new DockerTemplateOptions();
```
What is the purpose of having the options builder methods accepting a `null` value? Is there a need to "clear" any previously set value?

BTW, having a look at this class, it makes no sense to have the variables being an Optional. They're only used internally so they could be perfectly plain nullable objects. That would allow to cleanup the compute service adapter code and make it more readable. WDYT?

grkvlt replied:
The code is using the pattern found in other similar classes, see `SoftLayerTemplateOptions` and so on. If we're going to change this would it not make more sense to change _all_ of these in a separate pull request?

nacx replied:
A quick look to all existing template options subclasses show that the only ones that follow this pattern are the ones for Docker, SoftLayer and Azure Compute.
Given that this is not the common pattern and that this PR already cleans up the options in Docker, I think it's better to use this PR to remove the optionals from Docker, and fix SoftLayer and Azure in different PRs.

grkvlt replied:
OK, thanks @nacx I'll fix that here, and SoftLayer and Azure can be done elsewhere. I hadn't realised my choice of template option classes was so unusual!

nacx replied:
I wouldn't say unusual but "recent providers". I didn't notice this pattern was being used until I've had a deeper look at this PR. I'll take care of fixing SoftLayer and Azure. Thanks for spotting!

zack-shoylev commented on docker/src/main/java/org/jclouds/docker/compute/options/DockerTemplateOptions.java:
```diff
@@ -32,22 +34,24 @@
- * <code>
+ * Set<? extends NodeMetadata> set = api.createNodesInGroup(tag, 2, templateBuilder.build());}</pre>
  */
 public class DockerTemplateOptions extends TemplateOptions implements Cloneable {
```
It might be better to refactor this to autovalue and autovalue builders?

nacx replied:
Given that this extends the `TemplateOptions` class I'd say to leave this as-is, to avoid having a mixture of styles in one single class.
I agree this is something we can (and should) to in future PRs, but changing the TemplateOptions class in compute and all providers, but that is out of the scope of this PR.

zack-shoylev replied:
This is reasonable enough. Thanks @nacx

grkvlt commented:
@nacx @andreaturli I have changed the classes to remove the use of `Optional` and also updated the tests to reflect the change to `HostConfig`, is there anything else I need to do to make this available for merging with 1.9.1-SNAPSHOT as well as 2.0.0-SNAPSHOT?

andreaturli commented:
@grkvlt lgtm

@nacx, is it ok for you as well?

nacx commented on docker/src/main/java/org/jclouds/docker/compute/options/DockerTemplateOptions.java:
```diff
@@ -68,29 +71,29 @@ public void copyTo(TemplateOptions to) {
          }
-         if (hostname.isPresent()) {
-            eTo.hostname(hostname.get());
+         if (hostname != null) {
```
The null conditionals can be removed as these properties are `nullable` and there's no null check.

nacx commented on docker/src/main/java/org/jclouds/docker/compute/strategy/DockerComputeServiceAdapter.java:
```diff
@@ -97,38 +97,34 @@ public DockerComputeServiceAdapter(DockerApi api) {
 
-      if (templateOptions.getMemory().isPresent()) {
-         containerConfigBuilder.memory(templateOptions.getMemory().get());
+      if (templateOptions.getMemory() != null) {
```
Simplify this by removing the null check conditionals? These methods accept null values, so let's call them directly.

nacx commented:
Thanks @grkvlt! Just two minors, but lgtm. Mind addressing them while squashing the commits to prepare the merge?

grkvlt commented:
@nacx Thanks for the review, have squashed the commits and made the changes you suggested.

nacx commented:
Mind remove the conditionals [here](https://github.com/grkvlt/jclouds-labs/blob/fix/clocker/docker/src/main/java/org/jclouds/docker/compute/strategy/DockerComputeServiceAdapter.java#L104-L114) too?

grkvlt commented:
@nacx Good catch; done.

nacx commented:
Pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/7c5ded68) and [1.9.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/dcd1dc15). Thanks @grkvlt!
